README.md
## tcpsplice

![Chain Of Gophers](/chain.jpg?raw=true "Chain Of Gophers")

`tcpsplice` is an advanced TCP proxy/connection splicer written in Go.

## Requirements

## Build

## Configuration

## Monitoring

## Packaging


debian/changelog
tcpsplice (1.4.1) stable; urgency=medium

  * add a minimalist target load-balancing mechanism

 -- Pierre-Yves Kerembellec <py.kerembellec@gmail.com>  Sun, 19 Feb 2023 15:15:41 +0000

tcpsplice (1.4.0) stable; urgency=medium

  * add an in-band dead-hop bypass mechanism

 -- Pierre-Yves Kerembellec <py.kerembellec@gmail.com>  Fri, 17 Feb 2023 12:16:34 +0000

tcpsplice (1.3.2) stable; urgency=medium

  * you shall not try to outsmart thy TCP stack

 -- Pierre-Yves Kerembellec <py.kerembellec@gmail.com>  Fri, 04 Mar 2022 17:44:56 +0000

tcpsplice (1.3.1) stable; urgency=medium

  * fix crash condition

 -- Pierre-Yves Kerembellec <py.kerembellec@gmail.com>  Wed, 09 Jun 2021 18:07:01 +0200

tcpsplice (1.3.0) stable; urgency=medium

  * split code in different source files
  * revamp configuration file format
  * normalize logging format everywhere
  * add TLS mode to target connection
  * prepare for mTCP incoming connections
  * embed vendors static assets
  * use Bootstrap in monitor webui
  * add columns sort order in monitor webui
  * detect (and highlight) spliced incoming connections

 -- Pierre-Yves Kerembellec <py.kerembellec@gmail.com>  Thu, 20 May 2021 15:11:33 +0200

tcpsplice (1.2.0) stable; urgency=medium

  * add native TLS support and modernize software packaging

 -- Pierre-Yves Kerembellec <py.kerembellec@gmail.com>  Thu, 11 Feb 2021 16:45:09 +0100

tcpsplice (1.1.0) stable; urgency=medium

  * update to Go 1.15 (and Go modules)

 -- Pierre-Yves Kerembellec <py.kerembellec@gmail.com>  Mon, 21 Sep 2020 18:10:58 +0200

tcpsplice (1.0.2) stable; urgency=medium

  * add arbitrary metadata scan (+ display on monitoring screen)

 -- Pierre-Yves Kerembellec <py.kerembellec@gmail.com>  Mon, 26 Sep 2016 22:20:56 +0200

tcpsplice (1.0.1) stable; urgency=medium

  * initial version

 -- Pierre-Yves Kerembellec <py.kerembellec@gmail.com>  Wed, 09 Sep 2016 10:27:46 +0200


main.go
package main

import (
	"fmt"
	"os"
	"os/signal"
	"path/filepath"
	"syscall"
	"time"

	"github.com/pyke369/golang-support/uconfig"
	"github.com/pyke369/golang-support/ulog"
)

const PROGNAME = "tcpsplice"
const VERSION = "1.4.1"

var (
	Config  *uconfig.UConfig
	Logger  *ulog.ULog
	Started time.Time
	Unbind  bool
)

func main() {
	var err error

	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "usage: %s <configuration file>\n", filepath.Base(os.Args[0]))
		os.Exit(1)
	}
	if Config, err = uconfig.New(os.Args[1]); err != nil {
		fmt.Fprintf(os.Stderr, "configuration file syntax error: %s - aborting\n", err)
		os.Exit(2)
	}

	Logger = ulog.New(Config.GetString(PROGNAME+".log", "console(output=stdout)"))
	Logger.Info(map[string]any{
		"scope": "main", "event": "start", "version": VERSION, "config": os.Args[1],
		"pid": os.Getpid(), "services": len(Config.GetPaths(PROGNAME + ".service")),
	})
	Started = time.Now()

	go ServiceRun()
	go MonitorRun()

	signals := make(chan os.Signal, 1)
	signal.Notify(signals, syscall.SIGHUP, syscall.SIGUSR1)
	for {
		signal := <-signals
		switch {
		case signal == syscall.SIGHUP:
			if _, err = uconfig.New(os.Args[1]); err == nil {
				Config.Load(os.Args[1])
				Logger.Load(Config.GetString(PROGNAME+".log", "console(output=stdout)"))
				Logger.Info(map[string]any{
					"scope": "main", "event": "reload", "version": VERSION, "config": os.Args[1],
					"pid": os.Getpid(), "services": len(Config.GetPaths(PROGNAME + ".service")),
				})
			} else {
				Logger.Warn(map[string]any{
					"scope": "main", "event": "reload", "config": os.Args[1], "error": fmt.Sprintf("%v", err),
				})
			}

		case signal == syscall.SIGUSR1:
			Unbind = !Unbind
			if Unbind {
				Logger.Info(map[string]any{
					"scope": "main", "event": "unbind", "services": len(Config.GetPaths(PROGNAME + ".service")),
				})
			} else {
				Logger.Info(map[string]any{
					"scope": "main", "event": "rebind", "services": len(Config.GetPaths(PROGNAME + ".service")),
				})
			}
		}
	}
}


monitor.go
package main

import (
	"context"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"math"
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/pyke369/golang-support/acl"
	"github.com/pyke369/golang-support/dynacert"
)

var (
	MonitorUpdate   = make(chan *SESSION, 128)
	monitorSessions = map[string]*SESSION{}
	monitorLock     sync.RWMutex
)

func monitorHandle(response http.ResponseWriter, request *http.Request) {
	response.Header().Set("Server", PROGNAME+"/"+VERSION)
	response.Header().Set("Access-Control-Allow-Origin", "*")

	if request.Method == "OPTIONS" {
		return
	}
	if request.Method != "GET" {
		response.WriteHeader(http.StatusMethodNotAllowed)
		return
	}
	credentials, noauth := "", len(Config.GetPaths(PROGNAME+".monitor.noauth")) > 0 && acl.CIDRConfig(request.RemoteAddr, Config, PROGNAME+".monitor.noauth")
	if username, password, ok := request.BasicAuth(); ok {
		credentials = username + ":" + password
	}
	if !noauth && len(Config.GetPaths(PROGNAME+".monitor.auth")) > 0 && !acl.PasswordConfig(credentials, Config, PROGNAME+".monitor.auth") {
		response.Header().Set("WWW-Authenticate", fmt.Sprintf("Basic realm=\"%s\"", PROGNAME))
		response.WriteHeader(http.StatusUnauthorized)
		return
	}

	switch {
	case request.URL.Path == "/sessions.json":
		output := map[string]any{
			"server": map[string]any{
				"version": VERSION,
				"uptime":  time.Since(Started) / time.Second,
			},
		}
		services := map[string]any{}
		for _, name := range Config.GetPaths(PROGNAME + ".service") {
			name = strings.TrimPrefix(name, PROGNAME+".service.")
			entries := map[string]any{}
			monitorLock.RLock()
			for id, session := range monitorSessions {
				if session.name == name {
					entries[id] = map[string]any{
						"started":  session.started.Unix(),
						"duration": time.Since(session.started) / time.Second,
						"source":   session.source,
						"local":    session.local,
						"target":   session.target,
						"bytes":    [2]int64{session.smetrics.read, session.tmetrics.read},
						"mean":     [2]float64{session.smetrics.mean, session.tmetrics.mean},
						"last":     [2]float64{session.smetrics.last, session.tmetrics.last},
						"opaque":   session.opaque,
						"done":     session.done,
						"spliced":  session.spliced,
						"secure":   session.secure,
						"multi":    session.multi,
					}
				}
			}
			monitorLock.RUnlock()
			services[name] = entries
		}
		output["services"] = services
		response.Header().Set("Content-Type", "application/json")
		if json, err := json.Marshal(output); err == nil {
			response.Write(json)
		}

	case strings.HasPrefix(request.URL.Path, "/abort/"):
		monitorLock.RLock()
		for id, session := range monitorSessions {
			if id == request.URL.Path[7:] && !session.done {
				session.done = true
				session.abort <- true
				Logger.Warn(map[string]any{
					"scope": "monitor", "event": "abort", "id": id, "name": session.name,
					"source": session.source, "local": session.local, "target": session.target,
				})
				break
			}
		}
		monitorLock.RUnlock()
	}
}

func MonitorRun() {
	handler := http.NewServeMux()
	handler.HandleFunc("/sessions.json", monitorHandle)
	handler.HandleFunc("/abort/", monitorHandle)
	handler.Handle("/", http.StripPrefix("/", ResourcesHandler(6*time.Hour)))

	go func() {
		var server *http.Server

		key := ""
		for {
			time.Sleep(time.Second)
			if Unbind {
				if server != nil {
					Logger.Info(map[string]any{
						"scope": "monitor", "event": "close", "listen": strings.Split(key, "@@")[0],
					})
					server.Shutdown(context.Background())
					key, server = "", nil
				}
				continue
			}
			if parts := strings.Split(Config.GetStringMatch(PROGNAME+".monitor.listen", "_", `^.*?(:\d+)?((,[^,]+){2})?$`), ","); parts[0] != "_" {
				if value := strings.Join(parts, "@@"); value != key {
					if server != nil {
						Logger.Info(map[string]any{
							"scope": "monitor", "event": "close", "listen": strings.Split(key, "@@")[0],
						})
						server.Shutdown(context.Background())
					}
					key, server = value, &http.Server{
						Addr:         strings.TrimLeft(parts[0], "*"),
						Handler:      handler,
						ErrorLog:     log.New(io.Discard, "", 0),
						ReadTimeout:  10 * time.Second,
						WriteTimeout: 10 * time.Second,
						IdleTimeout:  30 * time.Second,
						TLSNextProto: map[string]func(*http.Server, *tls.Conn, http.Handler){},
					}
					if len(parts) > 1 {
						certificate := &dynacert.DYNACERT{}
						certificate.Add("*", parts[1], parts[2])
						server.TLSConfig = dynacert.IntermediateTLSConfig(certificate.GetCertificate)
						go func(server *http.Server, parts []string) {
							Logger.Info(map[string]any{
								"scope": "monitor", "event": "listen", "listen": parts[0], "certificate": strings.Join(parts[1:], ","),
							})
							server.ListenAndServeTLS("", "")
							key = ""
						}(server, parts)
					} else {
						go func(server *http.Server, parts []string) {
							Logger.Info(map[string]any{
								"scope": "monitor", "event": "listen", "listen": parts[0],
							})
							server.ListenAndServe()
							key = ""
						}(server, parts)
					}
				}
			}
		}
	}()

	go func() {
		for range time.Tick(time.Second) {
			count := 0
			monitorLock.Lock()
			for id, session := range monitorSessions {
				if time.Since(session.update) >= 3*time.Second {
					delete(monitorSessions, id)
					count++
				}
			}
			monitorLock.Unlock()
			if count > 0 {
				Logger.Info(map[string]any{
					"scope": "monitor", "event": "clean", "sessions": count,
				})
			}
		}
	}()

	for {
		session := <-MonitorUpdate
		id := session.id
		monitorLock.Lock()
		if monitorSessions[id] == nil {
			monitorSessions[id] = &SESSION{
				name:    session.name,
				source:  session.source,
				local:   session.local,
				target:  session.target,
				started: session.started,
				secure:  session.secure,
				multi:   session.multi,
				spliced: session.spliced,
				abort:   session.abort,
			}
		}
		duration := math.Max(float64(time.Since(session.started))/float64(time.Second), 0.001)
		session.smetrics.mean = math.Floor(float64(session.tmetrics.write*8)/(duration*1000)) / 1000
		session.tmetrics.mean = math.Floor(float64(session.smetrics.write*8)/(duration*1000)) / 1000
		if !session.update.IsZero() && !monitorSessions[id].update.IsZero() {
			duration = math.Max(float64(session.update.Sub(monitorSessions[id].update))/float64(time.Second), 0.001)
			session.smetrics.last = math.Floor(float64((session.tmetrics.write-monitorSessions[id].tmetrics.write)*8)/(duration*1000)) / 1000
			session.tmetrics.last = math.Floor(float64((session.smetrics.write-monitorSessions[id].smetrics.write)*8)/(duration*1000)) / 1000
		} else {
			session.smetrics.last = session.smetrics.mean
			session.tmetrics.last = session.tmetrics.mean
		}
		monitorSessions[id].opaque = session.opaque
		monitorSessions[id].update = session.update
		monitorSessions[id].done = session.done
		monitorSessions[id].smetrics.read = session.smetrics.read
		monitorSessions[id].smetrics.write = session.smetrics.write
		monitorSessions[id].smetrics.mean = session.smetrics.mean
		monitorSessions[id].smetrics.last = session.smetrics.last
		monitorSessions[id].tmetrics.read = session.tmetrics.read
		monitorSessions[id].tmetrics.write = session.tmetrics.write
		monitorSessions[id].tmetrics.mean = session.tmetrics.mean
		monitorSessions[id].tmetrics.last = session.tmetrics.last
		monitorLock.Unlock()
	}
}


tcpsplice.go
package main

import (
	"bytes"
	"crypto/md5"
	"encoding/json"
	"fmt"
	"github.com/pyke369/golang-support/uconfig"
	"github.com/pyke369/golang-support/ulog"
	"io/ioutil"
	"math"
	"math/rand"
	"net"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"time"
)

const progname = "tcpsplice"
const version = "1.0.3"

type Session struct {
	id, service, source, target, meta, targetName        string
	started, active, last                                time.Time
	sourceRead, targetWritten, targetRead, sourceWritten int64
	sourceMeanThroughput, targetMeanThroughput           float64
	sourceLastThroughput, targetLastThroughput           float64
	loggued, done                                        bool
	abort                                                chan bool
}

type Backend struct {
	Fqdn string
	Port int
}

var (
	config       *uconfig.UConfig
	logger       *ulog.ULog
	start        time.Time
	statistics   chan *Session
	sessions     map[string]*Session
	sessionsLock sync.RWMutex
	backendsURL  bytes.Buffer
)

func serviceHandler(service string, listener *net.TCPListener) {
	//Init service
	remotes := make([]string, 0)
	listenPort := listener.Addr().(*net.TCPAddr).Port
	backendUrl := config.GetString(fmt.Sprintf("services.%s.backend_url", service), "")
	static := true
	if strings.EqualFold(config.GetString(fmt.Sprintf("services.%s.remote_type", service), "static"), "dynamic") {
		logger.Info("%s Service : Dynamic backends are defined", service)
		//Initializing backends
		remotes = backendHandler(backendUrl, service, listenPort)
		if len(remotes) == 0 {
			logger.Warn("%s Service : Error while initializing Dynamic Backend, switching in static mode", service)
		} else {
			// Consider dynamic configuration is working so disabling static mode
			static = false
			backendTicker := time.NewTicker(time.Millisecond * 1000)
			go func() {
				for {
					select {
					case <-backendTicker.C:
						tmpRemotes := make([]string, 0)
						tmpRemotes = backendHandler(backendUrl, service, listenPort)
						if len(tmpRemotes) > 0 {
							remotes = tmpRemotes
						}
					}
				}
			}()
		}
	}
	if static {
		logger.Info("%s Service : Static backends will be used", service)
		remotes = config.GetPaths(fmt.Sprintf("services.%s.remote", service))
	}
	for {
		if source, err := listener.AcceptTCP(); err == nil {
			logger.Info("%s Service : Start Listening on port %v", service, listenPort)
			if len(remotes) > 0 {
				go func() {
					remote := remotes[rand.Int31n(int32(len(remotes)))]
					if target, err := net.DialTimeout("tcp", remote,
						time.Second*time.Duration(config.GetDurationBounds(fmt.Sprintf("services.%s.connect_timeout", service), 10, 2, 60))); err == nil {
						incomingSize := config.GetSizeBounds(fmt.Sprintf("services.%s.incoming_buffer_size", service), 64*1024, 4*1024, 512*1024)
						outgoingSize := config.GetSizeBounds(fmt.Sprintf("services.%s.outgoing_buffer_size", service), 64*1024, 4*1024, 512*1024)
						logMinimumSize := config.GetSizeBounds(fmt.Sprintf("services.%s.log_minimum_size", service), 1024, 0, math.MaxInt64)
						sessionMinimumSize := config.GetSizeBounds(fmt.Sprintf("services.%s.session_minimum_size", service), 1024, 0, math.MaxInt64)
						writeTimeout := time.Second * time.Duration(config.GetDurationBounds(fmt.Sprintf("services.%s.write_timeout", service), 10, 2, 60))
						idleTimeout := time.Second * time.Duration(config.GetDurationBounds(fmt.Sprintf("services.%s.idle_timeout", service), 60, 0, 300))
						metaSize := int(config.GetSizeBounds(fmt.Sprintf("services.%s.meta_size", service), 16*1024, 0, 64*1024))
						metaScan, _ := regexp.Compile(config.GetString(fmt.Sprintf("services.%s.meta_scan", service), "["))
						source.SetReadBuffer(int(incomingSize))
						source.SetWriteBuffer(int(incomingSize))
						target.(*net.TCPConn).SetReadBuffer(int(outgoingSize))
						target.(*net.TCPConn).SetWriteBuffer(int(outgoingSize))
						session := &Session{
							id:         fmt.Sprintf("%x", md5.Sum([]byte(fmt.Sprintf("%s%s", source.RemoteAddr(), target.RemoteAddr())))),
							service:    service,
							source:     fmt.Sprintf("%s", source.RemoteAddr()),
							target:     fmt.Sprintf("%s", target.RemoteAddr()),
							targetName: strings.Split(remote, ":")[0],
							started:    time.Now(),
							active:     time.Now(),
							last:       time.Time{},
							meta:       "-",
							abort:      make(chan bool, 1),
						}
						if logMinimumSize == 0 {
							session.loggued = true
							logger.Info(map[string]interface{}{"type": "splice", "id": session.id, "service": service, "source": session.source, "target": session.target, "targetName": session.targetName})
						}
						if sessionMinimumSize == 0 {
							statistics <- session
						}
						go func() {
							data := make([]byte, incomingSize)
							meta := make([]byte, 0, metaSize)
							for {
								read, err := source.Read(data)
								if read > 0 {
									if metaScan != nil && session.meta == "-" && len(meta) < metaSize {
										length := int(math.Min(float64(read), float64(metaSize-len(meta))))
										meta = append(meta, data[:length]...)
										if metaScan.Match(meta) {
											session.meta = ""
											for _, part := range metaScan.FindSubmatch(meta)[1:] {
												session.meta += string(part)
											}
											if session.meta == "" {
												session.meta = "-"
											}
										}
									}
									session.sourceRead += int64(read)
									session.active = time.Now()
									target.SetWriteDeadline(time.Now().Add(writeTimeout))
									written, err := target.Write(data[0:read])
									if written > 0 {
										session.targetWritten += int64(written)
									}
									if err != nil {
										logger.Warn(map[string]interface{}{"type": "error", "id": session.id, "service": service, "source": session.source, "target": session.target, "targetName": session.targetName, "error": "target write timeout"})
										break
									}
								}
								if err != nil {
									break
								}
							}
							source.Close()
							target.Close()
						}()
						data := make([]byte, outgoingSize)
						meta := make([]byte, 0, metaSize)
						close := false
						for !close {
							target.SetReadDeadline(time.Now().Add(time.Second))
							read, err := target.Read(data)
							if read > 0 {
								if metaScan != nil && session.meta == "-" && len(meta) < metaSize {
									length := int(math.Min(float64(read), float64(metaSize-len(meta))))
									meta = append(meta, data[:length]...)
									if metaScan.Match(meta) {
										session.meta = ""
										for _, part := range metaScan.FindSubmatch(meta)[1:] {
											session.meta += string(part)
										}
										if session.meta == "" {
											session.meta = "-"
										}
									}
								}
								session.targetRead += int64(read)
								session.active = time.Now()
								source.SetWriteDeadline(time.Now().Add(writeTimeout))
								written, err := source.Write(data[0:read])
								if written > 0 {
									session.sourceWritten += int64(written)
								}
								if err != nil {
									logger.Warn(map[string]interface{}{"type": "error", "id": session.id, "service": service, "source": session.source, "target": session.target, "error": "source write timeout"})
									break
								}
							}
							if err != nil {
								if err, ok := err.(net.Error); !ok || !err.Timeout() {
									break
								}
							}
							if !session.loggued && (session.targetWritten >= logMinimumSize || session.sourceWritten >= logMinimumSize) {
								session.loggued = true
								logger.Info(map[string]interface{}{"type": "splice", "id": session.id, "service": service, "source": session.source, "target": session.target, "targetName": session.targetName})
							}
							if (session.targetWritten >= sessionMinimumSize || session.sourceWritten >= sessionMinimumSize) && time.Now().Sub(session.last) >= time.Second*2 {
								session.last = time.Now()
								statistics <- session
							}
							select {
							case close = <-session.abort:
							default:
							}
							if idleTimeout != 0 && time.Now().Sub(session.active) >= idleTimeout {
								logger.Warn(map[string]interface{}{"type": "error", "id": session.id, "service": service, "source": session.source, "target": session.target, "error": "idle timeout"})
								close = true
							}
						}
						source.Close()
						target.Close()
						if session.targetWritten >= logMinimumSize || session.sourceWritten >= logMinimumSize {
							duration := math.Max(float64(time.Now().Sub(session.started))/float64(time.Second), 0.001)
							logger.Info(map[string]interface{}{"type": "unsplice", "id": session.id, "service": service, "source": session.source, "target": session.target,
								"duration": math.Floor(duration*1000) / 1000, "bytes": [2]int64{session.targetWritten, session.sourceWritten},
								"throughput": [2]float64{math.Floor(float64(session.targetWritten*8)/(duration*1000)) / 1000,
									math.Floor(float64(session.sourceWritten*8)/(duration*1000)) / 1000}})
						}
						if session.targetWritten >= sessionMinimumSize || session.sourceWritten >= sessionMinimumSize {
							session.done = true
							session.last = time.Now()
							statistics <- session
						}
					} else {
						source.Close()
					}
				}()
			} else {
				source.Close()
			}
		}
	}
}

func monitorHandler(response http.ResponseWriter, request *http.Request) {
	response.Header().Set("Access-Control-Allow-Origin", "*")
	if request.URL.Path == "/sessions.json" {
		output := map[string]interface{}{
			"server": map[string]interface{}{
				"version": version,
				"uptime":  time.Now().Sub(start) / time.Second,
			},
		}
		services := map[string]interface{}{}
		for _, service := range config.GetPaths("services") {
			service = strings.TrimPrefix(service, "services.")
			s := map[string]interface{}{}
			sessionsLock.RLock()
			for id, session := range sessions {
				s[id] = map[string]interface{}{
					"started":    session.started.Unix(),
					"duration":   time.Now().Sub(session.started) / time.Second,
					"source":     session.source,
					"target":     session.target,
					"targetName": session.targetName,
					"bytes":      [2]int64{session.sourceRead, session.targetRead},
					"mean":       [2]float64{session.sourceMeanThroughput, session.targetMeanThroughput},
					"last":       [2]float64{session.sourceLastThroughput, session.targetLastThroughput},
					"meta":       session.meta,
					"done":       session.done,
				}
			}
			sessionsLock.RUnlock()
			services[service] = s
		}
		output["services"] = services
		response.Header().Set("Content-Type", "application/json")
		if json, err := json.Marshal(output); err == nil {
			response.Write(json)
		}
	} else if strings.Index(request.URL.Path, "/abort/") == 0 {
		sessionsLock.RLock()
		for id, session := range sessions {
			if id == request.URL.Path[7:] && !session.done {
				session.done = true
				session.abort <- true
				logger.Warn(map[string]interface{}{"type": "abort", "id": id, "service": session.service, "source": session.source, "target": session.target})
				break
			}
		}
		sessionsLock.RUnlock()
	} else {
		response.Header().Set("Content-Type", "text/html; charset=utf-8")
		monitorFile, _ := ioutil.ReadFile("monitor.html")
		response.Write(monitorFile)
	}
}

func baseHandler(next http.Handler) http.Handler {
	return http.HandlerFunc(func(response http.ResponseWriter, request *http.Request) {
		response.Header().Set("Server", fmt.Sprintf("%s/%s", progname, version))
		if request.Method == "OPTIONS" {
			return
		}
		if request.Method != "HEAD" && request.Method != "GET" {
			response.WriteHeader(http.StatusMethodNotAllowed)
			return
		}
		check := false
		if remote, _, err := net.SplitHostPort(request.RemoteAddr); err == nil {
			if remote := net.ParseIP(remote); remote != nil {
				for _, allow := range config.GetPaths("monitor.acl.allow") {
					check = true
					allow = config.GetString(allow, "")
					if strings.Index(allow, "/") < 0 {
						allow += "/32"
					}
					if _, network, err := net.ParseCIDR(allow); err == nil {
						if remote != nil && network.Contains(remote) {
							next.ServeHTTP(response, request)
							return
						}
					}
				}
			}
		}
		credentials := ""
		if username, password, ok := request.BasicAuth(); ok {
			credentials = fmt.Sprintf("%s:%s", username, password)
		}
		for _, auth := range config.GetPaths("monitor.acl.auth") {
			check = true
			if config.GetString(auth, "") == credentials {
				next.ServeHTTP(response, request)
				return
			}
		}
		if check {
			response.Header().Set("WWW-Authenticate", fmt.Sprintf("Basic realm=\"%s\"", progname))
			response.WriteHeader(http.StatusUnauthorized)
			return
		}
		next.ServeHTTP(response, request)
	})
}

func backendHandler(backendUrl string, service string, port int) []string {
	backends := make([]Backend, 0)
	backendArray := make([]string, 0)
	//Makes HTTP request
	resp, err := http.Get(backendUrl)
	defer resp.Body.Close()
	if err != nil {
		logger.Warn("%s Service : %s", service, err)
		return backendArray
	}
	// Parses JSON response into
	err = json.NewDecoder(resp.Body).Decode(&backends)
	if err != nil {
		logger.Warn("%s Service : %s", service, err)
		return backendArray
	}
	// Converts Backend Array to String Array Containing URLs
	for _, backend := range backends {
		if backend.Port == 0 {
			backend.Port = port
		}
		var backendRTMPUrl bytes.Buffer
		backendRTMPUrl.WriteString(backend.Fqdn)
		backendRTMPUrl.WriteString(":")
		backendRTMPUrl.WriteString(strconv.Itoa(backend.Port))
		backendArray = append(backendArray, backendRTMPUrl.String())
	}
	return backendArray
}

func main() {
	var err error

	rand.Seed(time.Now().UnixNano() + int64(os.Getpid()))
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "usage: %s <configuration file>\n", filepath.Base(os.Args[0]))
		os.Exit(1)
	}
	if config, err = uconfig.New(os.Args[1]); err != nil {
		fmt.Fprintf(os.Stderr, "configuration file syntax error: %s - aborting\n", err)
		os.Exit(2)
	}
	logger = ulog.New(config.GetString("server.log", "console()"))
	logger.Info(map[string]interface{}{"type": "start", "version": version, "config": os.Args[1], "services": len(config.GetPaths("services"))})
	start = time.Now()

	for _, service := range config.GetPaths("services") {
		for _, local := range config.GetPaths(fmt.Sprintf("%s.local", service)) {
			service = strings.TrimPrefix(service, "services.")
			local = config.GetString(local, "")
			if address, err := net.ResolveTCPAddr("tcp", strings.TrimLeft(local, "*")); err == nil {
				if listener, err := net.ListenTCP("tcp", address); err == nil {
					logger.Info(map[string]interface{}{"type": "service", "name": service, "listen": local})
					go serviceHandler(service, listener)
				} else {
					logger.Warn(map[string]interface{}{"type": "service", "name": service, "listen": local, "error": err})
				}
			} else {
				logger.Warn(map[string]interface{}{"type": "service", "name": service, "listen": local, "error": err})
			}
		}
	}

	http.Handle("/", baseHandler(http.HandlerFunc(monitorHandler)))
	if parts := strings.Split(config.GetStringMatch("monitor.listen", "_", "^(?:\\*|\\d+(?:\\.\\d+){3}|\\[[^\\]]+\\])(?::\\d+)?(?:(?:,[^,]+){2})?$"), ","); parts[0] != "_" {
		server := &http.Server{Addr: strings.TrimLeft(parts[0], "*")}
		if len(parts) > 1 {
			logger.Info(map[string]interface{}{"type": "monitor", "listen": parts[0], "cert": parts[1], "key": parts[2]})
			go server.ListenAndServeTLS(parts[1], parts[2])
		} else {
			logger.Info(map[string]interface{}{"type": "monitor", "listen": parts[0]})
			go server.ListenAndServe()
		}
	}

	statistics = make(chan *Session, 100)
	sessions = map[string]*Session{}
	cleaner := time.NewTicker(time.Second).C
	for {
		select {
		case session := <-statistics:
			sessionsLock.Lock()
			if sessions[session.id] == nil {
				sessions[session.id] = &Session{service: session.service, source: session.source, target: session.target, targetName: session.targetName, started: session.started, abort: session.abort}
			}
			sessionsLock.Unlock()
			duration := math.Max(float64(time.Now().Sub(session.started))/float64(time.Second), 0.001)
			session.sourceMeanThroughput = math.Floor(float64(session.targetWritten*8)/(duration*1000)) / 1000
			session.targetMeanThroughput = math.Floor(float64(session.sourceWritten*8)/(duration*1000)) / 1000
			if !session.last.IsZero() && !sessions[session.id].last.IsZero() {
				duration = math.Max(float64(session.last.Sub(sessions[session.id].last))/float64(time.Second), 0.001)
				session.sourceLastThroughput = math.Floor(float64((session.targetWritten-sessions[session.id].targetWritten)*8)/(duration*1000)) / 1000
				session.targetLastThroughput = math.Floor(float64((session.sourceWritten-sessions[session.id].sourceWritten)*8)/(duration*1000)) / 1000
			} else {
				session.sourceLastThroughput = session.sourceMeanThroughput
				session.targetLastThroughput = session.targetMeanThroughput
			}
			sessions[session.id].last = session.last
			sessions[session.id].done = session.done
			sessions[session.id].meta = session.meta
			sessions[session.id].sourceRead = session.sourceRead
			sessions[session.id].targetWritten = session.targetWritten
			sessions[session.id].targetRead = session.targetRead
			sessions[session.id].sourceWritten = session.sourceWritten
			sessions[session.id].sourceMeanThroughput = session.sourceMeanThroughput
			sessions[session.id].targetMeanThroughput = session.targetMeanThroughput
			sessions[session.id].sourceLastThroughput = session.sourceLastThroughput
			sessions[session.id].targetLastThroughput = session.targetLastThroughput
		case <-cleaner:
			var ids []string
			sessionsLock.RLock()
			for id, session := range sessions {
				if time.Now().Sub(session.last) >= time.Second*3 {
					ids = append(ids, id)
				}
			}
			sessionsLock.RUnlock()
			sessionsLock.Lock()
			for _, id := range ids {
				delete(sessions, id)
			}
			sessionsLock.Unlock()
			if len(ids) > 0 {
				logger.Info(map[string]interface{}{"type": "clean", "sessions": len(ids)})
			}
		}
	}
}
